cpp/src/io/orc/writer_impl.cu

```diff
@@ -1121,96 +1147,163 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
+      auto const stat_end   = stat_begin + rowgroup_merge[i].num_chunks;
       rowgroup_blobs[i].assign(stat_begin, stat_end);
     }
     return rowgroup_blobs;
   }();
 
+  return {std::move(rowgroup_blobs),
+          std::move(stripe_chunks),
+          std::move(stripe_merge),
+          std::move(col_stats_dtypes),
+          std::move(col_types)};
+}
+
+writer::impl::encoded_footer_statistics writer::impl::finish_statistic_blobs(
+  int num_stripes, writer::impl::persisted_statistics& per_chunk_stats)
+{
+  auto stripe_size_iter = thrust::make_transform_iterator(per_chunk_stats.stripe_stat_merge.begin(),
+                                                          [](auto const& i) { return i.size(); });
+
+  auto const num_columns = per_chunk_stats.col_types.size();
+  auto const num_stripe_blobs =
+    thrust::reduce(stripe_size_iter, stripe_size_iter + per_chunk_stats.stripe_stat_merge.size());
+  auto const num_file_blobs = num_columns;
+  auto const num_blobs = single_write_mode ? static_cast<int>(num_stripe_blobs + num_file_blobs)
+                                           : static_cast<int>(num_stripe_blobs);
+
+  if (num_stripe_blobs == 0) { return {}; }
+
+  // merge the stripe persisted data and add file data
+  rmm::device_uvector<statistics_chunk> stat_chunks(num_blobs, stream);
+  hostdevice_vector<statistics_merge_group> stats_merge(num_blobs, stream);
+
+  size_t chunk_offset = 0;
+  size_t merge_offset = 0;
+  for (size_t i = 0; i < per_chunk_stats.stripe_stat_chunks.size(); ++i) {
+    auto chunk_bytes = per_chunk_stats.stripe_stat_chunks[i].size() * sizeof(statistics_chunk);
+    auto merge_bytes = per_chunk_stats.stripe_stat_merge[i].size() * sizeof(statistics_merge_group);
+    cudaMemcpyAsync(stat_chunks.data() + chunk_offset,
+                    per_chunk_stats.stripe_stat_chunks[i].data(),
+                    chunk_bytes,
+                    cudaMemcpyDeviceToDevice,
+                    stream);
+    cudaMemcpyAsync(stats_merge.device_ptr() + merge_offset,
+                    per_chunk_stats.stripe_stat_merge[i].device_ptr(),
+                    merge_bytes,
+                    cudaMemcpyDeviceToDevice,
+                    stream);
+    chunk_offset += per_chunk_stats.stripe_stat_chunks[i].size();
+    merge_offset += per_chunk_stats.stripe_stat_merge[i].size();
+  }
+
+  if (single_write_mode) {
+    std::vector<statistics_merge_group> file_stats_merge(num_file_blobs);
+    for (auto i = 0u; i < num_file_blobs; ++i) {
+      auto col_stats         = &file_stats_merge[i];
+      col_stats->col_dtype   = per_chunk_stats.col_types[i];
+      col_stats->stats_dtype = per_chunk_stats.stats_dtypes[i];
+      col_stats->start_chunk = static_cast<uint32_t>(i * num_stripes);
+      col_stats->num_chunks  = static_cast<uint32_t>(num_stripes);
+    }
+
+    auto d_file_stats_merge = stats_merge.device_ptr(num_stripe_blobs);
+    cudaMemcpyAsync(d_file_stats_merge,
```

Comment:
It looks like this is using a raw `cudaMemcpyAsync` on a `hostdevice_vector` object. Can we use its built-in methods to avoid a raw `cudaMemcpyAsync` call?

Reply:
Nice catch!